Delete failed payment attempts for successfully settled payments

tvolk131 opened the pull request:
Fixes #6318

Failed payment attempts are deleted from channel.db for successfully settled payments, unless the `keep-failed-payment-attempts` flag is specified in configuration.

Summarizing part of the discussion below:
We only want to delete failed attempts for _settled_ payments. If we were to delete failed attempts for failed payments, that would leave all failed payments as empty shells, so we might as well delete the whole payment. This might be a nice optional feature to have, but should be handled in a separate PR (if at all). And we don't want to delete failed attempts for inflight payments since inflight payments could end up failing.

bhandras commented:
Thanks for the PR @tvolk131 :+1: 
Please fix the commit messages (reference: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure)
Quick example for one line messages:
"My commit" => "package: my commit"

Also please make sure that checks don't fail (at least one `itest` should also succeed).

bhandras commented on routing/control_tower_test.go:
```diff
@@ -174,21 +183,31 @@ func TestControlTowerSubscribeSuccess(t *testing.T) {
 }
 
-func testPaymentControlSubscribeFail(t *testing.T, registerAttempt bool) {
+func testPaymentControlSubscribeFail(t *testing.T, registerAttempt, keepFailedPayments bool) {
```
We have some formatting rules that are non-standard to `gofmt`, please check: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-spacing-and-formatting

tvolk131 replied:
Done.

bhandras commented on channeldb/payment_control.go:
```diff
@@ -95,12 +95,20 @@ type PaymentControl struct {
+// PaymentControlConfig houses necessary fields that a PaymentControl needs to
+// function.
+type PaymentControlConfig struct {
+	KeepFailedPayments bool
```
nit: add godoc since this member is exported.

tvolk131 replied:
Done.

bhandras commented on channeldb/payment_control.go:
```diff
@@ -538,6 +546,15 @@ func (p *PaymentControl) Fail(paymentHash lntypes.Hash,
 
+		if !p.cfg.KeepFailedPayments &&
+			payment.FailureReason != nil &&
+			payment.Status == StatusFailed {
```
The payment may still have some active shards, so it's quite likely that calling `deletePaymentTx` will fail here (see `routing/payment_lifecycle.go`)

tvolk131 replied:
I took a look at `routing/payment_lifecycle.go` as you suggested. However, `fetchPayment` in `channeldb/payments.go`, which constructs the MPPayment instance, ensures that payment.Status will only be set to StatusFailed if there are no active shards for the payment. I also realized that the `payment.FailureReason != nil` check I had there was redundant since the Fail function always sets the FailureReason to a non-nil value.

tvolk131 replied:
Resolving this, since I have changed the PR to the point where this conversation is no longer relevant.

tvolk131 commented:
Dang, I thanks for the speedy turnaround on this @bhandras! Looks like the only non-trivial issue is of deleting the payment while some payment shards are still in-flight. I'm hoping to add a test case for this and turn it back to you late tonight, but it might end up being more like tomorrow afternoon.

Also, I apologize for the number of style violations I made... I'm a new contributor and am still getting used to the codebase and style guide. Appreciate you bearing with me :)

bhandras commented:
> Dang, I thanks for the speedy turnaround on this @bhandras! Looks like the only non-trivial issue is of deleting the payment while some payment shards are still in-flight. I'm hoping to add a test case for this and turn it back to you late tonight, but it might end up being more like tomorrow afternoon.
> 

Yeah, the payment lifecycle is really non-trivial and takes some time to fully understand. No need to rush, take your time :)

> Also, I apologize for the number of style violations I made... I'm a new contributor and am still getting used to the codebase and style guide. Appreciate you bearing with me :)

No worries. These are just some light rules to make the code look roughly the same readability wise.

tvolk131 commented on server.go:
```diff
@@ -885,7 +885,11 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+		KeepFailedPayments: cfg.KeepFailedPayments,
+	}
+
+	paymentControl := channeldb.NewPaymentControl(dbs.ChanStateDB, paymentControlConfig)
```
Done.

tvolk131 commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -231,6 +231,8 @@ then watch it on chain. Taproot script spends are also supported through the
 * Add [htlc expiry protection](https://github.com/lightningnetwork/lnd/pull/6212)
 to the htlc interceptor API.
 
+* [Delete failed payments](https://github.com/lightningnetwork/lnd/pull/6438)
```
Done.

tvolk131 commented on channeldb/payments.go:
```diff
@@ -734,87 +734,94 @@ func (d *DB) DeletePayment(paymentHash lntypes.Hash,
+// of an existing transaction.
+func (d *DB) deletePaymentTx(paymentHash lntypes.Hash,
+	failedHtlcsOnly bool, tx kvdb.RwTx) error {
+	payments := tx.ReadWriteBucket(paymentsRootBucket)
```
Done.

tvolk131 commented:
@bhandras I addressed all of your comments - however, I don't know how to retroactively change the messages for previous git commits. I didn't see anything in the contribution guide on how to do that, and everything I saw on Google looked like wizard-level git trickery. Is there any resource on how to do this?

tvolk131 commented:
Bump! Anything else that needs to be done on my end?

bhandras commented:
> @bhandras I addressed all of your comments - however, I don't know how to retroactively change the messages for previous git commits. I didn't see anything in the contribution guide on how to do that, and everything I saw on Google looked like wizard-level git trickery. Is there any resource on how to do this?

You can try interactive rebase to change history. In this case something like `git rebase -i HEAD~5`. Then `r` means reword, `s` means squash a commit to a prev one, `e` to edit that commit, `p` simply to pick as is etc. Once  commit is ready you can `git rebase --continue`. Maybe if you're not familiar try it on another test branch to get the hang of it.

bhandras commented:
gg

> Bump! Anything else that needs to be done on my end?

Please re-request (in the reviewers list above) when ready for re-review. This notifies reviewers that the PR is ready for another round.

yyforyongyu commented on routing/control_tower_test.go:
```diff
@@ -237,16 +256,28 @@ func testPaymentControlSubscribeFail(t *testing.T, registerAttempt bool) {
+	} else {
+		// Cannot register a second subscriber after the payment has failed.
+		_, err := pControl.SubscribePayment(info.PaymentIdentifier)
+		assert.Error(t, err)
```
There are mixed usage of `t.Fatalf`, `assert` and `require` in this commit. Moving forward, we want to use `require` only.

tvolk131 replied:
Fixed. However, for tests which are direct copies of existing tests such as `channeldb/payment_control_test.go`, I am continuing to match whatever assertions are in the existing test - feel free to unresolve this if you think I should change this.

yyforyongyu commented on channeldb/payment_control.go:
```diff
@@ -95,12 +95,20 @@ type PaymentControl struct {
 
 // NewPaymentControl creates a new instance of the PaymentControl.
-func NewPaymentControl(db *DB) *PaymentControl {
+func NewPaymentControl(db *DB, cfg *PaymentControlConfig) *PaymentControl {
```
The config probably doesn't belong here. Please check `config_builder.go` re [how we set options for db](https://github.com/lightningnetwork/lnd/blob/master/config_builder.go#L850). Since it's db-related config, it makes sense to add the config option there, and access it via the `db` instance here.

tvolk131 replied:
Fixed!

yyforyongyu reviewed (changes requested):
Thank you for the PR! I like that a great amount of effort has been taken into writing the tests! And I've also noticed you've been very active in this repo and want you to know it's very welcomed! However, there are a few basic aspects before we spend time to take a serious look into the PR, specifically for this PR,
- the git message doesn't summarize the changes made inside that commit
- the git message doesn't have the required format
- code is not formatted as required.

Besides, the original issue is to delete failed attempts when possible, which means it's not only for failed payments as we may have lots of failed attempts but the payment still succeeds.

tvolk131 commented:
> Please re-request (in the reviewers list above) when ready for re-review. This notifies reviewers that the PR is ready for another round.

Gotcha, sorry about that! Will do.

tvolk131 commented:
> Thank you for the PR! I like that a great amount of effort has been taken into writing the tests! And I've also noticed you've been very active in this repo and want you to know it's very welcomed!

Thanks! It's awesome to be able to contribute to such a meaningful project.

> However, there are a few basic aspects before we'd spend time to take a serious look into the PR, specifically for this PR,
> 
> * the git message doesn't summarize the changes made inside that commit
> * the git message doesn't have the required format
> * code is not formatted as required.

Sorry, I should have paid closer attention to the contribution guide! I'll be giving that a very thorough read, and I appreciate y'all bearing with me.

> Besides, the original issue is to delete failed attempts when possible, which means it's not only for failed payments as we may have lots of failed attempts but the payment still succeeds.

Ahh I see that now. I can have a PR for that out by tomorrow (implementation was quick, I just need to add some tests). Is this current PR of any value, or should I just scrap it?

yyforyongyu commented:
> Ahh I see that now. I can have a PR for that out by tomorrow (implementation was quick, I just need to add some tests). Is this current PR of any value, or should I just scrap it?

I'd suggest keeping this PR and revising/polishing it as needed since @bhandras and I have already started the review process(maybe also a good opportunity for you to practice `git rebase`🤓).  It is expected to have multiple review rounds before a PR can be accepted, so please be mentally prepared!

And a few thoughts re deleting failed attempts, with the new flag,
- we should delete them for payments that are in the terminal state - failed/settled. 
- we may need to have a db migration for the old payments created prior to this new flag?

Also the build needs to pass except for the itests as there are known issues in them.

tvolk131 commented:
Apologies for the delay on this! Implementation is done, as mentioned previously, I just want to shore up some more test coverage before sending back for review.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -957,6 +927,78 @@ func (d *DB) DeletePayments(failedOnly, failedHtlcsOnly bool) error {
 
+// deletePaymentAttempt removes an HTLC attempt from disk and returns the
+// payment in the state before it was deleted.
+func (d *DB) deletePaymentAttempt(hash lntypes.Hash, attemptID uint64) (*MPPayment, error) {
```
nit: line too long (exceeds the 80 limit)

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -779,24 +780,7 @@ func (d *DB) deletePaymentTx(paymentHash lntypes.Hash,
-				htlcBucketKey(htlcSettleInfoKey, htlcID),
-			)
-			if err != nil {
+			if err = deletePaymentAttemptTx(htlcsBucket, htlcID); err != nil {
```
nit: line too long

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -918,21 +902,7 @@ func (d *DB) DeletePayments(failedOnly, failedHtlcsOnly bool) error {
-				if err := htlcsBucket.Delete(
-					htlcBucketKey(htlcSettleInfoKey, aid),
-				); err != nil {
+				if err = deletePaymentAttemptTx(htlcsBucket, aid); err != nil {
```
nit: line too long

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -957,6 +927,78 @@ func (d *DB) DeletePayments(failedOnly, failedHtlcsOnly bool) error {
+		payment = nil
+		payments := tx.ReadWriteBucket(paymentsRootBucket)
+		if payments == nil {
+			return nil
```
I think we need to return an error here and below if buckets cannot be found.

tvolk131 replied:
Done. And I updated the function header to reflect this.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -957,6 +927,78 @@ func (d *DB) DeletePayments(failedOnly, failedHtlcsOnly bool) error {
+		htlcsBucket := bucket.NestedReadWriteBucket(
+			paymentHtlcsBucket,
+		)
+
```
very nit: unnecessary new line

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payments.go:
```diff
@@ -957,6 +927,78 @@ func (d *DB) DeletePayments(failedOnly, failedHtlcsOnly bool) error {
+		p, err := fetchPayment(bucket)
+		payment = p
+
+		if err != nil {
```
Usually the pattern is, we check the returned `err` first before everything else, so it's better to check `err != nil` before the above assignment.

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/db.go:
```diff
@@ -315,6 +317,12 @@ func (d *DB) Path() string {
 
+// SetKeepFailedPayments sets whether failed payments should be deleted.
+// It is only used for testing.
+func (d *DB) SetKeepFailedPayments(keepFailedPayments bool) {
```
Can we just use `OptionKeepFailedPayments`?

tvolk131 replied:
Good point. Done.

yyforyongyu commented on channeldb/options.go:
```diff
@@ -61,6 +61,8 @@ type Options struct {
 	// database if set to true.
 	dryRun bool
+
+	keepFailedPayments bool
```
missing doc

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payment_control_test.go:
```diff
@@ -196,6 +197,155 @@ func TestPaymentControlSwitchFail(t *testing.T) {
+	db, cleanup, err := MakeTestDB()
+	defer cleanup()
+	if err != nil {
+		t.Fatalf("unable to init db: %v", err)
```
nit: use `require.NoError`, same below.

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payment_control_test.go:
```diff
@@ -196,6 +197,155 @@ func TestPaymentControlSwitchFail(t *testing.T) {
+	// Lookup the payment so we can get its old sequence number before it is
+	// overwritten.
+	payment, err := pControl.FetchPayment(info.PaymentIdentifier)
+	assert.NoError(t, err)
```
nit: use `require` instead

tvolk131 replied:
Done.

I used `assert.NoError` here because another test in this file also was. There are a handful of `assert.NoError` instances that must have slipped through in the past. I added a commit to this PR that changes them.

yyforyongyu replied:
Yeah I saw that one, very nice PR btw! We did have a mixed usage of both, and decided to stick to `require` but never had the time to clean the old ones, so thanks for that PR!

yyforyongyu commented on channeldb/payments_test.go:
```diff
@@ -635,6 +635,51 @@ func TestFetchPaymentWithSequenceNumber(t *testing.T) {
+	// Deleting payment attempt when no payments
+	// exist should not return error.
+	payment, err := db.deletePaymentAttempt(
+		info.PaymentIdentifier, attempt.AttemptID)
```
From contribution guidelines,

WRONG
```
value, err := bar(a,
	a, b, c)
```
RIGHT
```
value, err := bar(
	a, a, b, c,
)
```

tvolk131 replied:
Done.

yyforyongyu reviewed (changes requested):
What about these two?
> And a few thoughts re deleting failed attempts, with the new flag,
> - we should delete them for payments that are in the terminal state - failed/settled.
> - we may need to have a db migration for the old payments created prior to this new flag?

From the original issue description,
> Only keep around failed payment attempts if specified by a new flag. 

Maybe a bit confusing, but "failed payments" and "failed payment attempts" are two different things.

tvolk131 commented on channeldb/mp_payment.go:
```diff
@@ -224,10 +224,10 @@ func (m *MPPayment) InFlightHTLCs() []HTLCAttempt {
 func (m *MPPayment) GetAttempt(id uint64) (*HTLCAttempt, error) {
-	for _, htlc := range m.HTLCs {
-		htlc := htlc
+	for i := range m.HTLCs {
```
Without this change, one of the unit tests in the `routing` package fails:

```
$ make unit pkg=routing
 Running unit tests.
go test  -tags="dev nolog"  -test.timeout=60m github.com/lightningnetwork/lnd/routing
--- FAIL: TestPaymentControlSubscribeFail (0.07s)
    --- FAIL: TestPaymentControlSubscribeFail/register_attempt,_delete_failed_payments (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7abd56]

goroutine 4869 [running]:
testing.tRunner.func1.2({0x87a340, 0xd4cbb0})
        /home/gitpod/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
        /home/gitpod/go/src/testing/testing.go:1392 +0x39f
panic({0x87a340, 0xd4cbb0})
        /home/gitpod/go/src/runtime/panic.go:838 +0x207
github.com/lightningnetwork/lnd/routing.testPaymentControlSubscribeFail(0xc00b0dad00, 0x1, 0x0)
        /workspace/lnd/routing/control_tower_test.go:240 +0x4f6
github.com/lightningnetwork/lnd/routing.TestPaymentControlSubscribeFail.func2(0xc000503520?)
        /workspace/lnd/routing/control_tower_test.go:182 +0x25
testing.tRunner(0xc00b0dad00, 0xa2cbf8)
        /home/gitpod/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
        /home/gitpod/go/src/testing/testing.go:1486 +0x35f
FAIL    github.com/lightningnetwork/lnd/routing 0.346s
```

In the previous version of GetAttempt, `htlc` as defined in the for loop is a copy, *not a reference*. That's important because we want PaymentControl.FailAttempt to return the payment object. If the config is set to keep failed payment attempts, FailAttempt writes the failure info to disk and then re-fetches and returns the payment attempt, which works fine. However, if the config is set to delete failed attempts, it reads the attempt from disk, wipes it from disk, modifies it in memory, and then returns it. The in-memory modification is where this problem arises, since we're calling GetAttempt to get a reference to the htlc and append the failure info. And if you look at where the test is failing:

`if *htlcAttempt.Failure != failInfo {`

We're attempting to dereference the failure, which is nil because we appended it to a copy of the htlc.

tvolk131 replied:
Just added a test in `channeldb/mp_payment_test.go` (TestGetAttemptReturnsReferenceToPayment) that directly tests this behavior.

tvolk131 commented:
@yyforyongyu just want to get some clarification on what you mentioned a few days ago...

> What about these two?
> 
> > And a few thoughts re deleting failed attempts, with the new flag,
> > 
> > * we should delete them for payments that are in the terminal state - failed/settled.
> > * we may need to have a db migration for the old payments created prior to this new flag?

Are you suggesting that:
1. Rather than immediately deleting failed attempts, we should write them to disk. And later, when the payment settles, we iterate through all attempts within the payment and delete the ones that failed.
2. We want to delete failed attempts from disk right when they fail, but in addition, we want a DB migration that sifts through old payments and cleans up failed attempts.

> Maybe a bit confusing, but "failed payments" and "failed payment attempts" are two different things.

I fully understand the conceptual difference between a payment and a payment attempt. I just got a little tripped up while getting familiar with the codebase and accidentally implemented deletion of failed payments rather than failed attempts. Currently this PR does both - that is, it deletes failed payment attempts *and* failed payments. Should I remove the code that deletes failed payments? Or is that something we want as well?

tvolk131 commented:
@yyforyongyu just requested re-review! I have one question about migration. I added migration code that removes all failed payment attempts for previous payments. However, I wasn't quite sure what the most idiomatic way is to read from the config when performing the migration to potentially skip the deletion of these attempts if `keepFailedPayments` is set. My first thought would be to have `syncVersions`, which has access to `DB.keepFailedPayments`, call a new optional parameter on each version in the `versions` argument which allows the potential to skip that migration. That's kinda a one-off solution thoug, feels a little bit hacky since migrations almost always introduce breaking changes, and therefore should _never_ be skipped.

Another note about the new migration code - I didn't bother adding the ability for this migration to remove failed _payments_, only payment attempts, since I'm still not certain on whether we even want to remove failed payments. If we do, I will update the migration code to match that behavior.

yyforyongyu reviewed (changes requested):
> That's kinda a one-off solution though, and feels a little bit hacky since migrations almost always introduce breaking changes, and therefore should never be skipped.

Good thought! Tbh migration is hard, not to mention we are introducing an optional migration here. We are building a new framework to do the [optional migration here](https://github.com/lightningnetwork/lnd/pull/6469). So I'm thinking maybe we could move the migration into a new PR since we'd rebase that PR.

> Currently this PR does both - that is, it deletes failed payment attempts and failed payments. Should I remove the code that deletes failed payments? Or is that something we want as well?

Yeah we only want to delete failed attempts, not the payments as the users may want them for accounting purposes. The main reason we want to delete failed attempts is that, for successful payments, we may end up having lots of failed attempts since we use MPP here and not keeping them can save us some space.

> Rather than immediately deleting failed attempts, we should write them to disk. And later, when the payment settles, we iterate through all attempts within the payment and delete the ones that failed.

Right on point. Meanwhile I'd suggest going through the payment lifecycle to grasp the relationship of a payment and its htlc attempts. iirc, a payment relies on its attempts to determine its state, which means we cannot delete the attempts while the payment is still in-flight, that suggests we can only safely delete the attempts when the payment is in the terminal state - either settled or failed.

However, for a failed payment, all of its attempts must have failed too. Thus auto-deleting the attempts here would make little sense(I think) since the failed payment would then be left with no attempt info and makes debugging difficult. Plus the user could just use `DeletePayment`.

tvolk131 commented on blockcache/blockcache_test.go:
```diff
@@ -173,10 +172,10 @@ func TestBlockCacheMutexes(t *testing.T) {
 			if e%2 == 0 {
 				_, err := bc.GetBlock(&blockhash1, getBlockImpl)
-				assert.NoError(t, err)
+				require.NoError(t, err)
```
Are you talking #6504? That fixes something similar but slightly different. I can definitely move this commit to a new PR though!

yyforyongyu replied:
Yeah exactly, seems like it belongs there. Plus it'd be easier for the reviewers to focus on payment related changes in this PR to shorten the review cycle.

tvolk131 replied:
Moved to #6636.

tvolk131 commented:
> Yeah we only want to delete failed attempts, not the payments as the users may want them for accounting purposes.

Not trying to push back on this point, but I'm curious what the accounting purpose would be for failed payments. Is it because a MPP can still have shards that individually succeed? In that case, we _could_ delete AMPs that have failed, or MPPs that have failed _and_ have no successful shards, right? Not that we need to address that in this PR, I'm just trying to wrap my head around it.

Roasbeef commented:
> Not trying to push back on this point, but I'm curious what the accounting purpose would be for failed payments.

Still need to catch up with the above thread, and this PR, but was wondering this myself. The only case I can think of is service wanting to keep track of what % of payments are failed _entirely_ as a sort of health/success metric. In that case, I think that lines up with the proposed `keep-failed-payments` flag, since those that need that information can opt into retaining it.

yyforyongyu commented:
> Not trying to push back on this point, but I'm curious what the accounting purpose would be for failed payments. Is it because a MPP can still have shards that individually succeed? 

If any of the shards succeeds then the payment won't be failed. If the payment is failed then all its shards must have failed.

Imo it's the same reason to keep the settled payments. You make a payment, and you want to later query it to get some info. It's like the history of your transactions I think? Maybe I'm overthinking here, but I think people tend to pay attention when things go wrong. That's why I think we'd better keep the failed attempts for failed payments so it'd be easier to investigate, and they can be deleted via rpc later on. The `keep-failed-payments` can then take effect on settled payments to save space. Plus I assume most of the payments will be settled so keeping the failed payments won't grow the disk too much.

We could also delete failed attempts for failed payments too, then not that much info is left and the failed payments are more or less empty shells. In that case, we might as well delete the payments themselves too. Maybe this could be accomplished using a different config flag so the operators can easily distinguish and have finer control.

tvolk131 commented:
What about a `keep-failed-payment-attempts` flag to explicitly keep failed attempts, and a `delete-failed-payments` flag to explicitly delete failed payments?

Also, I want to get feedback on my design approach before I move forward now that I realize that we don't want to wipe failed payment attempts from disk until _after_ the payment settles, rather than immediately upon attempt failure. If I understand correctly, an HTLC could still return as failed _after_ another HTLC for the same payment has settled. If that's the case, then whenever any HTLC attempt is marked as settled _or_ failed, we should delete all failed HTLCs if there are no more in-flight HTLCs for that payment, and at least one of the HTLCs was settled. Here's some quick pseudocode to help clarify:

```
func (p *PaymentControl) SettleAttempt(hash lntypes.Hash,
	attemptID uint64, settleInfo *HTLCSettleInfo) (*MPPayment, error) {

        // Marks as settled in db.
        // ...

        if containsNoInflightPaymentAttempts(hash) {
                deleteFailedPaymentAttempts(hash)
        }

        // ...
}

func (p *PaymentControl) FailAttempt(hash lntypes.Hash,
	attemptID uint64, failInfo *HTLCFailInfo) (*MPPayment, error) {

        // Marks as settled in db.
        // ...

        if containsNoInflightPaymentAttempts(hash) && containsSettledHTLC(hash) {
                deleteFailedPaymentAttempts(hash)
        }

        // ...
}
```

lightninglabs-deploy commented:
@bhandras: review reminder
@tvolk131, remember to re-request review from reviewers when ready

yyforyongyu reviewed (commented):
> an HTLC could still return as failed after another HTLC for the same payment has settled. 

Correct. Generally it's better not to assume message ordering in a p2p environment. So the set of HTLCs won't come back in order. However, this will be very unlikely as the receiver would be sending us free money if it happens. The receiver will make sure only after the last HTLC is received then the HTLCs are settled.

When deciding on payment state, I think it's better to work on payment level, not attempt level. So the deletion needs to happen inside the `lifecycle`, after we know the payment is terminated, such as `currentState.terminated()`. 

> What about a keep-failed-payment-attempts flag to explicitly keep failed attempts, and a delete-failed-payments flag to explicitly delete failed payments?

As for the flags, let's focus on attempts removal first?

lightninglabs-deploy commented:
@bhandras: review reminder
@tvolk131, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@tvolk131, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@tvolk131, remember to re-request review from reviewers when ready

tvolk131 commented on channeldb/options.go:
```diff
@@ -61,6 +61,10 @@ type Options struct {
 	// database if set to true.
 	dryRun bool
+
+	// keepFailedPayments determines whether failed htlc attempts are kept
```
Done.

tvolk131 commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -231,6 +231,9 @@ then watch it on chain. Taproot script spends are also supported through the
 * Add [htlc expiry protection](https://github.com/lightningnetwork/lnd/pull/6212)
 to the htlc interceptor API.
 
+* [Delete failed payments](https://github.com/lightningnetwork/lnd/pull/6438)
```
Done.

tvolk131 commented:
@yyforyongyu sorry for ghosting on this! With travel and family stuff, I let this fall by the wayside... But I'm back!

> When deciding on payment state, I think it's better to work on payment level, not attempt level. So the deletion needs to happen inside the `lifecycle`, after we know the payment is terminated, such as `currentState.terminated()`.

That makes sense to work on the higher-level abstraction of payments instead of payment attempts, since we want to delete the failed attempts when the _payment_ is finished, not when an attempt succeeds. However I would still argue that `PaymentControl.SettleAttempt` is a better place to delete payment attempts. The current description for the `paymentLifecycle` struct (which is what calls `currentState.terminated()`) is:

```paymentLifecycle holds all information about the current state of a payment needed to resume if from any point```

Deleting failed payment attempts doesn't really relate to the state of a payment - it's more of a persistence detail. Whereas the description for the PaymentControl struct is:

```PaymentControl implements persistence for payments and payment attempts```

I would argue that deleting failed payment attempts falls much more closely into this definition.

Just my 2 cents - I'm happy to hear any pushback on this point!

tvolk131 commented:
@yyforyongyu I _radically_ simplified this PR from what it was originally. And after digging into the code more, I understand what you were talking about with the reference to `currentState.terminated()`. In the interest of holding each struct to their single responsibility, I added a new method (`CleanupPayment`) to the `ControlTower` interface which is called at the payment level and allows the control tower to perform any final persistence-level cleanup actions after the payment lifecycle decides that a payment is in a terminal state.

I also moved the migration code over to another branch, which I'll open a PR for once this PR is merged.

yyforyongyu commented on docs/release-notes/release-notes-0.16.0.md:
```diff
@@ -0,0 +1,10 @@
+
+## Database
+
+* [Delete failed payment attempts](https://github.com/lightningnetwork/lnd/pull/6438)
```
Could aim for `v0.15.1` first!

tvolk131 commented on channeldb/payment_control_test.go:
```diff
@@ -1003,6 +1003,73 @@ func TestPaymentControlMPPRecordValidation(t *testing.T) {
+	// Cleaning up a successful payment should remove failed htlcs if
+	// keepFailedPaymentAttempts is not enabled.
+	require.NoError(t, pControl.CleanupPayment(payments[2].id))
+	if !keepFailedPaymentAttempts {
```
Done. See comment.

tvolk131 commented on channeldb/payment_control.go:
```diff
@@ -223,6 +223,25 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 	return updateErr
 }
 
+// CleanupPayment deletes all failed htlcs
```
Done.

tvolk131 commented on routing/control_tower.go:
```diff
@@ -125,6 +132,12 @@ func (p *controlTower) InitPayment(paymentHash lntypes.Hash,
 	return p.db.InitPayment(paymentHash, info)
 }
 
+// CleanupPayment deletes all failed htlcs
```
Done.

tvolk131 commented on routing/payment_lifecycle.go:
```diff
@@ -183,6 +183,11 @@ lifecycle:
 		// We have a terminal condition and no active shards, we are
 		// ready to exit.
 		case currentState.terminated():
+			err := p.router.cfg.Control.CleanupPayment(p.identifier)
```
Ahh good catch! Done.

tvolk131 commented on channeldb/payment_control_test.go:
```diff
@@ -1003,6 +1003,73 @@ func TestPaymentControlMPPRecordValidation(t *testing.T) {
+	db, cleanup, err := MakeTestDB()
+	db.keepFailedPaymentAttempts = keepFailedPaymentAttempts
+	defer cleanup()
+	require.NoError(t, err, "unable to init db")
```
Fixed!

tvolk131 commented on routing/control_tower.go:
```diff
@@ -19,6 +19,13 @@ type ControlTower interface {
+	// htlcs are completed, and the payment has reached a final state.
+	// This means that it should only be called after either SettleAttempt
+	// or FailAttempt is called on the final htlc.
+	CleanupPayment(lntypes.Hash) error
```
I updated the documentation for that method, LMK if you think it still needs some reworking.

tvolk131 commented on channeldb/payment_control_test.go:
```diff
@@ -1003,6 +1003,73 @@ func TestPaymentControlMPPRecordValidation(t *testing.T) {
+
+	pControl := NewPaymentControl(db)
+
+	// Register three payments:
```
Thanks! It's mostly copied from another existing test

tvolk131 commented:
> I'm working on a PR to radically change how we manage the payment state

> maybe you could help us review that PR🤪

Absolutely! Could you link it here for reference, or shoot a link to me on Slack whenever it's ready?

bhandras commented on routing/control_tower_test.go:
```diff
@@ -286,13 +294,13 @@ func testPaymentControlSubscribeFail(t *testing.T, registerAttempt bool) {
 	}
 
-	db, err := channeldb.Open(tempPath)
+	db, err := channeldb.Open(tempPath, channeldb.OptionKeepFailedPaymentAttempts(keepFailedPaymentAttempts))
```
nit: formatting

tvolk131 replied:
Fixed.

tvolk131 replied:
Done.

bhandras commented on channeldb/payment_control.go:
```diff
@@ -223,6 +223,17 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+// DeleteFailedAttempts deletes all failed htlcs for a payment.
+func (p *PaymentControl) DeleteFailedAttempts(hash lntypes.Hash) error {
+	if !p.db.keepFailedPaymentAttempts {
+		err := p.db.DeletePayment(hash, true)
```
nit: maybe worth adding a const here to name what true means in this context (for readability).

tvolk131 replied:
Done - at least I think. Did I do what you were asking me to do?

bhandras commented on routing/control_tower.go:
```diff
@@ -19,6 +19,11 @@ type ControlTower interface {
 	// hash.
 	InitPayment(lntypes.Hash, *channeldb.PaymentCreationInfo) error
 
+	// DeleteFailedAttempts removes all failed HTLCs from disk. It should be
```
nit: from disk/from the db

tvolk131 replied:
Done.

bhandras commented on routing/payment_lifecycle.go:
```diff
@@ -187,6 +187,11 @@ lifecycle:
 			// the preimage and route.
 			for _, a := range payment.HTLCs {
 				if a.Settle != nil {
+					err := p.router.cfg.Control.DeleteFailedAttempts(p.identifier)
```
Given that we have a settled payment here, isn't that a problem that we do fail the lifecycle if the delete fails for whatever reason? Perhaps we could log and error or critical instead? @yyforyongyu

yyforyongyu replied:
Very good point! Yeah we dare not fail the payment here since it's settled. Just to be safe we should log an error here instead.

tvolk131 replied:
Ahh good catch, that makes sense! Done.

yyforyongyu replied:
hmmm did you forget to push the latest update?

bhandras reviewed (approved):
LGTM, very nice and clean PR, great work @tvolk131 :tada: 
I've added a few nits and one question we may want to discuss.

yyforyongyu commented on channeldb/payment_control.go:
```diff
@@ -223,6 +223,18 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 	return updateErr
 }
 
+// DeleteFailedAttempts deletes all failed htlcs for a payment.
```
nit: maybe update docs saying this would only happen if the db config is set?

tvolk131 replied:
Done.

yyforyongyu commented on channeldb/payment_control.go:
```diff
@@ -223,6 +223,18 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 
+// DeleteFailedAttempts deletes all failed htlcs for a payment.
+func (p *PaymentControl) DeleteFailedAttempts(hash lntypes.Hash) error {
+	if !p.db.keepFailedPaymentAttempts {
```
Non-blocking, just something to think about. I think originally we have the flag `keepFailedPaymentAttempts` placed in `channeldb` rather than in `routing` because we want to introduce an optional migration that deletes the attempts, which makes total sense as it's db's responsibility to manage the migrations. On the other hand, it's nice to have "pure" CRUD methods in `channeldb` and have constraints made at the callsite to maximally generalize those CRUD methods. In our case, `DeleteFailedAttempts` can just do its job without checking the flag, and let `routing` decide whether to prune the HTLCs based on its own config, which means we'd have another flag placed in `routing` for that purpose. A big however is, since we already have the flag in db, it can perfectly serve the two purposes so we'd rather only have one flag there. I'll continue the onion-removal work on payments and invoices to see how it fit the whole picture. Atm, I think your approach is the best.

tvolk131 replied:
I'm a bit unclear on the plan going forward. Are you saying we could move the flag from `channeldb` to `routing` after the optional migration is added?

yyforyongyu reviewed (changes requested):
Very niiiiice! Just need to address @bhandras 's comment re logging the error and we'd be good to go!

tvolk131 commented:
Just did a fresh rebase! I believe this is ready to submit.

Roasbeef commented:
Looks like a syntax error was introduced during the latest rebase: 
> Error: routing/payment_lifecycle.go:195:25: syntax error: unexpected newline, expecting comma or )

tvolk131 commented:
Fixed!

Roasbeef commented on channeldb/options.go:
```diff
@@ -164,3 +168,9 @@ func OptionDryRunMigration(dryRun bool) OptionModifier {
 	}
 }
+
+func OptionKeepFailedPaymentAttempts(keepFailedPaymentAttempts bool) OptionModifier {
```
Style nit: missing a `godoc` comment.

tvolk131 replied:
Done!

Roasbeef commented on channeldb/payment_control.go:
```diff
@@ -223,6 +223,19 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+// by the PaymentControl db.
+func (p *PaymentControl) DeleteFailedAttempts(hash lntypes.Hash) error {
+	if !p.db.keepFailedPaymentAttempts {
+		const failedHtlcsOnly = true
```
Makes me wish Go had named arguments like Python, but I like the style of introducing some intermediate variables when it makes sense to make the semantics of the passed arguments clearer. It def beats: `foo(1, true, false, true)`.

tvolk131 replied:
> It def beats: foo(1, true, false, true).

Yeah for sure 😆

Roasbeef commented on channeldb/payment_control_test.go:
```diff
@@ -1003,6 +1003,94 @@ func TestPaymentControlMPPRecordValidation(t *testing.T) {
+	// Calling DeleteFailedAttempts on a failed payment should delete all
+	// HTLCs.
+	require.NoError(t, pControl.DeleteFailedAttempts(payments[0].id))
+	// Expect all HTLCs to be deleted if the config is set to delete them.
```
Style nit: missing newline.

tvolk131 replied:
Done.

Roasbeef commented on channeldb/payment_control_test.go:
```diff
@@ -1003,6 +1003,94 @@ func TestPaymentControlMPPRecordValidation(t *testing.T) {
+	} else {
+		require.Error(t, pControl.DeleteFailedAttempts(payments[1].id))
+	}
+	// Since DeleteFailedAttempts returned an error, we should expect the
```
Style nit: missing newline.

tvolk131 replied:
Done.

Roasbeef commented:
> Summarizing part of the discussion below:


Thanks for this summary! I fell off of this PR for a bit, but this helped me go back and trace the evolution of the PR. Sometimes we set out with an initial target, but then later realize either that target isn't really feasible, or may violate some existing invariants in the system.

Roasbeef reviewed (approved):
LGTM 🎨

Thanks for being so patient and responsive with this PR 🙏🏾

guggero commented on config.go:
```diff
@@ -199,6 +199,10 @@ const (
 	// used by default to fund transactions.
 	defaultCoinSelectionStrategy = "largest"
+
+	// defaultKeepFailedPaymentAttempts is the default setting for whether to keep
```
nit: line too long. https://github.com/lightningnetwork/lnd/pull/6681 should help devs make this easier to avoid by auto-setting displayed tab with to 8 spaces and adding a ruler at 80 chars.

tvolk131 replied:
Done.

guggero commented on config.go:
```diff
@@ -625,6 +631,7 @@ func DefaultConfig() Config {
 		PendingCommitInterval:   defaultPendingCommitInterval,
 		ChannelCommitBatchSize:  defaultChannelCommitBatchSize,
 		CoinSelectionStrategy:   defaultCoinSelectionStrategy,
+		KeepFailedPaymentAttempts:      defaultKeepFailedPaymentAttempts,
```
Looks like this file should be formatted... Not sure why the linter didn't pick this up.

tvolk131 replied:
Done. Not sure how that slipped by...

guggero commented on routing/control_tower_test.go:
```diff
@@ -300,7 +300,9 @@ func initDB(keepFailedPaymentAttempts bool) (*channeldb.DB, error) {
 	}
 
-	db, err := channeldb.Open(tempPath, channeldb.OptionKeepFailedPaymentAttempts(keepFailedPaymentAttempts))
+	db, err := channeldb.Open(tempPath,
```
This should be fixed in the previous commit.
And the correct formatting would be:
```
	db, err := channeldb.Open(
		tempPath, channeldb.OptionKeepFailedPaymentAttempts(
			keepFailedPaymentAttempts,
		).
	)
```

tvolk131 replied:
Done.

guggero commented on routing/payment_lifecycle.go:
```diff
@@ -187,6 +187,21 @@ lifecycle:
 			// the preimage and route.
 			for _, a := range payment.HTLCs {
 				if a.Settle != nil {
+					err := p.router.cfg.Control.
```
How about:

```
			// Find the first successful shard and return
			// the preimage and route.
			for _, a := range payment.HTLCs {
				if a.Settle == nil {
					continue
				}

				err := p.router.cfg.Control.DeleteFailedAttempts(
					p.identifier,
				)
				if err != nil {
					log.Errorf("Error deleting failed "+
						"payment attempts for payment "+
						"%v: %v", p.identifier, err)
				}

				return a.Settle.Preimage, &a.Route, nil
			}
```
?

tvolk131 replied:
I like it! Done.

guggero reviewed (commented):
Wanted to hit the merge button, but found a few nits of my own while scanning the code :sweat_smile:
